src/pipeline-stages/src/main/scala/TextPreprocessor.scala

```diff
@@ -0,0 +1,157 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import breeze.linalg.max
+import org.apache.spark.ml.Transformer
+import org.apache.spark.ml.param.{MapParam, Param, ParamMap}
+import org.apache.spark.ml.util.{ComplexParamsReadable, ComplexParamsWritable, Identifiable}
+import org.apache.spark.sql.types.{StringType, StructField, StructType}
+import org.apache.spark.sql.{DataFrame, Dataset}
+import org.apache.spark.sql.functions.udf
+import spray.json.DefaultJsonProtocol._
+
+class Trie(trieMap: Map[String, Trie] = Map[String, Trie](),
+           textValue: String = "",
+           normFunc: String => String = (text:String) => text.toLowerCase()) extends Serializable {
+  val value: String = textValue
+
+  def getNodeValue(key: String, value:String): String = (key, value) match {
+    case (k, v) if k.length() == 1 => v
+    case (_, _) => ""
+  }
+
+  def getNode(key: String, value:String): Trie = {
+    val newLeaf = new Trie(textValue = getNodeValue(key, value), normFunc = this.normFunc)
+    this.trieMap.getOrElse(key.substring(0, 1), newLeaf)
+  }
+
+  def put(key:String, value:String, isNormalized:Boolean = false): Trie = (key, value) match {
+    case (_, _) if key.length() == 0 => this.copy()
+    case (k, v) =>
+      var normalizedKey:String = k
+      if (!isNormalized) normalizedKey = this.normFunc(k)
```

Comment:
really try to avoid using `var` in general, in this case, you would make this

`val normalizedKey = if (isNormalized) k else normFunk(k)`

Reply:
Good call on one lining. However, I think "this." is useful since parameters could be similarly named. Is there any strong opinion on this from the team?